File: m3core/src/m3rt/base/toolbox.cpp
/* 
M3 -- Meka Robotics Real-Time Control System
Copyright (c) 2010 Meka Robotics
Author: edsinger@mekabot.com (Aaron Edsinger)

M3 is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

M3 is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with M3.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <m3rt/base/toolbox.h>
#include <stdarg.h>
#include <string.h>
#include <stdlib.h> 
#include <stdio.h>
#include <math.h>
#include <iostream>
#include <stdarg.h>
#include <cstdlib>
#include <time.h>

namespace m3rt
{
 using namespace std;
  
 void M3_WARN(const char * format, ...)
 {
	FILE * pFile = NULL;
	string path;
	
	if (GetEnvironmentVar(M3_ROBOT_ENV_VAR, path))
	{		
		path.append(LOG_FILE);
		pFile = fopen (path.c_str(),"a");
	}	
	
	 char buffer[256];
	 va_list args;
	 va_start (args, format);
	 vsprintf (buffer, format, args);
	 printf("M3 WARNING: ");
	 printf("%s", buffer);
	 if (pFile)
	 {
	    fprintf(pFile,"M3 WARNING: ");
	    fprintf(pFile,"%s", buffer);
	    fclose (pFile);
	 }
	 va_end (args);
 }

 void M3_ERR(const char * format, ...)
 {
	FILE * pFile = NULL;
	string path;
	
	if (GetEnvironmentVar(M3_ROBOT_ENV_VAR, path))
	{		
		path.append(LOG_FILE);
		pFile = fopen (path.c_str(),"a");
	}	
   
	 char buffer[256];
	 va_list args;
	 va_start (args, format);
	 vsprintf (buffer, format, args);
	 printf("M3 ERROR: ");
	 printf("%s", buffer);
	 if (pFile)
	 {
	    fprintf(pFile,"M3 ERROR: ");
	    fprintf(pFile,"%s", buffer);
	    fclose (pFile);
	 }
	 va_end (args);
 }
 
  void M3_DEBUG(const char * format, ...)
 {
	 char buffer[256];
	 va_list args;
	 va_start (args, format);
	 vsprintf (buffer, format, args);
	 printf("M3 DEBUG: ");
	 printf("%s", buffer);
	 va_end (args);
 }

 void M3_INFO(const char * format, ...)
 {
	FILE * pFile = NULL;
	string path;
	
	if (GetEnvironmentVar(M3_ROBOT_ENV_VAR, path))
	{		
		path.append(LOG_FILE);
		pFile = fopen (path.c_str(),"a");
	}	
   
	 char buffer[256];
	 va_list args;
	 va_start (args, format);
	 vsprintf (buffer, format, args);
	 printf("M3 INFO: ");
	 printf("%s", buffer);
	 if (pFile)
	 {
	    fprintf(pFile,"M3 INFO: ");
	    fprintf(pFile,"%s", buffer);
	    fclose (pFile);
	 }
	 va_end (args);
 }
 
void BannerPrint(int width, const char *format, ...)
{
	char buf[81];
	width=width>80?80:width;
	char out[]="--------------------------------------------------------------------------------"; //len 80
	va_list args;
	va_start (args, format);
	snprintf(buf,80,format,args);
	M3_PRINTF("\n");
	int n=(int)strlen(buf);
	if (n<width&&((width-n)/2)>0)
	{
		out[(width-n)/2]=0;
		M3_PRINTF(out);
		M3_PRINTF(buf);
		M3_PRINTF(out);
	}
	else
		M3_PRINTF(buf);
	va_end (args);
	M3_PRINTF("\n");
}

bool GetEnvironmentVar(const char * var, string &s)
{
	char *p=getenv(var);
	if (p!=NULL)
	{
		s.assign(p);
		return true;
	}
	return false;
}

vector<mReal> YamlReadVectorM(string s)
{
	size_t start=s.find_first_of( "[",0);//, size_type num );
	size_t end=s.find_first_of( ",]",0);
	vector<mReal> f;
	while (start<s.size()&&end<s.size()&&start!=string::npos&&end!=string::npos)
	{
		string val=s.substr(start+1,end-start-1);
		f.push_back((mReal) strtod(val.c_str(), NULL ));
		start=end;
		end=s.find_first_of( ",]", start+1);
	}
	return f;
};

vector<string> YamlReadVectorString(string s)
{
	size_t start=s.find_first_of( "[",0);//, size_type num );
	size_t end=s.find_first_of( ",]",0);
	vector<string> f;
	while (start<s.size()&&end<s.size()&&start!=string::npos&&end!=string::npos)
	{
		string val=s.substr(start+1,end-start-1);
		f.push_back(val);
		start=end;
		end=s.find_first_of( ",]", start+1);
	}
	return f;
};

void operator >> (const YAML::Node& node, vector<mReal> & v)
{
	for(unsigned i=0;i<node.size();i++) {
		mReal x;
		node[i] >> x;
		v.push_back(x);
	}
}

vector<mReal> YamlReadVectorM(const YAML::Node& seq)
{	
	vector<mReal> f;
	mReal val;
	for (int i=0; i<seq.size(); i++)
	{
		seq[i] >> val;
		f.push_back(val);
	}

	return f;
};

// Converts a hexadecimal string to integer
// Returns 0 if not valid
unsigned int xtoi(const char* xs)
{
	unsigned int n;
	sscanf(xs,"%x",&n);
	return n;
}

void WriteYamlDoc(const char * filename, YAML::Emitter &doc)
{
	string s(filename);
	string path;
	if (GetEnvironmentVar(M3_ROBOT_ENV_VAR, path))
	{		
		path.append("/robot_config/");
		path.append(s);
	}
	std::ofstream fout(path.c_str());
	fout << doc.c_str();
	fout.close();
}

void GetYamlDoc(const char * filename, YAML::Node & doc)
{	
	string s(filename);
	string path;
	
	if (GetEnvironmentVar(M3_ROBOT_ENV_VAR, path))
	{		
		path.append("/robot_config/");
		path.append(s);
	}
	ifstream fin(path.c_str());
	if (fin.fail())
	{		
		M3_ERR("could not read %s \n", path.c_str());	
	}

   	YAML::Parser parser(fin);
   	
   	parser.GetNextDocument(doc);
	fin.close();
	return;
}

}

File: m3meka/python/m3/toolbox.py
#M3 -- Meka Robotics Robot Components
#Copyright (c) 2010 Meka Robotics
#Author: edsinger@mekabot.com (Aaron Edsinger)

#M3 is free software: you can redistribute it and/or modify
#it under the terms of the GNU Lesser General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#M3 is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU Lesser General Public License for more details.

#You should have received a copy of the GNU Lesser General Public License
#along with M3.  If not, see <http://www.gnu.org/licenses/>.
#import matplotlib
#matplotlib.use('TkAgg')
import pylab

import os
import sys
import time
import exceptions
import yaml
import Gnuplot
import numpy as nu
#import Numeric
import glob
import datetime
from datetime import timedelta
from m3.unit_conversion import *
from threading import Thread
import m3.component_base_pb2 as mbs
#import roslib; roslib.load_manifest('m3_toolbox_ros')
#import rospy
#import rosbag
from m3.toolbox_core import *


def get_m3_animation_path():
        try:
                return os.environ['M3_ROBOT']+'/animations/'
        except KeyError:
                print 'SET YOUR M3_ROBOT ENVIRONMENT VARIABLE'
        return ''


def get_omnibase_pwr_component_name(name):
        config= None	
        try:
                f=file(get_component_config_filename(name),'r')
                config= yaml.safe_load(f.read())
        except (IOError, EOFError):
                print 'Config file not present:',get_component_config_filename(name),'for',name
                return
        jnt_array_name=config['joint_array_component']
        return get_chain_pwr_component_name(jnt_array_name)

def get_chain_pwr_component_name(name):		
        names=get_chain_joint_names(name)
        if len(names)>0:
                return get_joint_pwr_component_name(names[0])
        else:
                return ''

def get_joint_pwr_component_name(name):
        return get_actuator_pwr_component_name(get_joint_actuator_component_name(name))

def get_actuator_pwr_component_name(name):
        return get_actuator_ec_pwr_component_name(get_actuator_ec_component_name(name))

def get_actuator_ec_pwr_component_name(name):
        config= None	
        try:
                f=file(get_component_config_filename(name),'r')
                config= yaml.safe_load(f.read())
        except (IOError, EOFError):
                print 'Config file not present:',get_component_config_filename(name),'for',name
                return
        return config['pwr_component']

def get_joint_actuator_component_name(name):	
        if get_component_config_type(name) == 'm3joint_slave':
                return ""
        config= None	
        try:
                f=file(get_component_config_filename(name),'r')
                config= yaml.safe_load(f.read())
        except (IOError, EOFError):
                print 'Config file not present:',get_component_config_filename(name),'for',name
                return
        return config['actuator_component']
        
def get_joint_ctrl_component_name(name):	
        if get_component_config_type(name) == 'm3joint_slave':
                return ""
        config= None	
        try:
                f=file(get_component_config_filename(name),'r')
                config= yaml.safe_load(f.read())
        except:
                print 'Config file not present:',get_component_config_filename(name),'for',name
                return ""
        ctrl = ""
        try:
		ctrl = config['control_component']
	except:
		ctrl = ""
        return ctrl


def get_chain_dynamatics_component_name(name):		
        config= None	
        try:
                f=file(get_component_config_filename(name),'r')
                config= yaml.safe_load(f.read())
        except (IOError, EOFError):
                print 'Config file not present:',get_component_config_filename(name),'for',name
                return

        if config.has_key('dynamatics_component'):
                return config['dynamatics_component']
        else:
                return ''

def get_chain_loadx6_name(chain_name):
        if (chain_name.find('arm')==-1):
                print 'Error: get_chain_loadx6_name() valid for arms only'
                return None
        joint_names=get_chain_joint_names(chain_name)
        j0=joint_names[0]
        return 'm3loadx6_'+j0[j0.find('_')+1:j0.find('_',j0.find('_')+1)]+'_l0'

def get_chain_limb_name(name):		
        config= None	
        try:
                f=file(get_component_config_filename(name),'r')
                config= yaml.safe_load(f.read())
        except (IOError, EOFError):
                print 'Config file not present:',get_component_config_filename(name),'for',name
                return
        if config.has_key('limb_name'):
                return config['limb_name']
        else:
                return None

def get_chain_joint_names(name):		
        config= None	
        try:
                f=file(get_component_config_filename(name),'r')
                config= yaml.safe_load(f.read())
        except (IOError, EOFError):
                print 'Config file not present:',get_component_config_filename(name),'for',name
                return
        if config.has_key('joint_components'):
                #Have to sort keys by *_JID*'
                ret=[]
                for jid in range(len(config['joint_components'])):
                        ret.append(config['joint_components']['J'+str(jid)])
                return ret
        else:
                return []

def get_chain_joint_limits(name):
        l=[]
        names=get_chain_joint_names(name)
        for n in names:
                try:
                        f=file(get_component_config_filename(n),'r')
                        config= yaml.safe_load(f.read())
                        l.append([config['param']['min_q'],config['param']['max_q']])
                except (IOError, EOFError):
                        print 'Config file not present:',get_component_config_filename(name),'for',name
                        return []
        return l

def get_actuator_ec_component_name(name):
        config= get_component_config(name)	
        if config is not None:
                return config['ec_component']
        else:
                return None

def get_joint_chain_name(name):
        path=get_m3_config_path()
        filename= path+'m3_config.yml'
        f=file(filename,'r')
        config= yaml.safe_load(f.read())

        for cdir in config['rt_components'].keys():
                for c in config['rt_components'][cdir].keys():
                        if (c==name):
                                for d in config['rt_components'][cdir].keys():
                                        if config['rt_components'][cdir][d] == 'm3arm' or \
                                           config['rt_components'][cdir][d] == 'm3torso' or \
                                           config['rt_components'][cdir][d] == 'm3hand' or \
                                           config['rt_components'][cdir][d] == 'm3head':
                                                return d
        return ''

def get_robot_name():
        path=get_m3_config_path()
        filename= path+'m3_config.yml'
        f=file(filename,'r')
        config= yaml.safe_load(f.read())

        for cdir in config['rt_components'].keys():
                for c in config['rt_components'][cdir].keys():
                        if config['rt_components'][cdir][c] == 'm3humanoid' or config['rt_components'][cdir][c] == 'm3humanoid_sea':
                                return c
        return ''
        return ''

"""
* Animation file reader for files provided by Andrea Thomaz
* each line is a trajectory via point: ts q0 q1 ... qn qd0 qd1 ... qdn qdd0 qdd1 ... qddn
* n dof (7 each arm + torso: n=17) (define joint ordering in the comments)
* ts: timestamp (s)
* qx: joint angle of dof x (rad)
* qdx: joint velocity of dof x (rad/s)
* qddx: joint accel of dof x (rad/s2)

Note: the joint angles seem scale by a factor of 0.5, and the velocities are all >0
       we'll just use the joint angles anyhow
"""

def get_animation_files():
        path=get_m3_animation_path()
        return glob.glob(path+'*.txt')

def get_animation_names():
        path=get_m3_animation_path()
        full=glob.glob(path+'*.txt')
        names=[]
        for s in full:
                names.append(s[s.rfind('/')+1:s.rfind('.')])
        return names

def load_animation(filename):
        f=open(filename,'rb')
        data=f.read()
        nl=data.find('\r\n')
        samples=[]
        while nl>=0:
                sample=[]
                line=data[:nl]
                ts=line.find('\t')
                while ts>=0:
                        sample.append(float(line[:ts]))
                        line=line[ts+1:]
                        ts=line.find('\t')
                sample.append(float(line))
                samples.append(sample)
                data=data[nl+1:]
                nl=data.find('\r\n')
        x=nu.array(samples)

        #limb names should map to chain limb names
        animation={'timestamp':x[:,0],
                   'torso':{    'theta':x[:,1:4],'thetadot':x[:,18:21],'thetadotdot':x[:,35:38]},
                   'left_arm':{'theta':-2*x[:,4:11], 'thetadot':x[:,21:28],'thetadotdot':x[:,38:45]},
                   'right_arm':{'theta':-2*x[:,11:18], 'thetadot':x[:,28:35],'thetadotdot':x[:,45:52]}}

        ##patch animation to work with meka bodies
        animation['torso']['theta'][:,1:3]=animation['torso']['theta'][:,1:2]*-1
        animation['torso']['thetadot'][:,1:3]=animation['torso']['thetadot'][:,1:2]*-1
        animation['torso']['thetadotdot'][:,1:3]=animation['torso']['thetadotdot'][:,1:2]*-1

        animation['right_arm']['theta'][:,2]=animation['right_arm']['theta'][:,2]*-1
        animation['right_arm']['thetadot'][:,2]=animation['right_arm']['thetadot'][:,2]*-1
        animation['right_arm']['thetadotdot'][:,2]=animation['right_arm']['thetadotdot'][:,2]*-1

        animation['left_arm']['theta'][:,2]=animation['left_arm']['theta'][:,2]*-1
        animation['left_arm']['thetadot'][:,2]=animation['left_arm']['thetadot'][:,2]*-1
        animation['left_arm']['thetadotdot'][:,2]=animation['left_arm']['thetadotdot'][:,2]*-1

        animation['left_arm']['theta'][:,1]=animation['left_arm']['theta'][:,1]*-1
        animation['left_arm']['thetadot'][:,1]=animation['left_arm']['thetadot'][:,1]*-1
        animation['left_arm']['thetadotdot'][:,1]=animation['left_arm']['thetadotdot'][:,1]*-1

        #animation['left_arm']['theta'][:,0]=animation['left_arm']['theta'][:,0]*-1
        #animation['left_arm']['thetadot'][:,0]=animation['left_arm']['thetadot'][:,0]*-1
        #animation['left_arm']['thetadotdot'][:,0]=animation['left_arm']['thetadotdot'][:,0]*-1
        return animation

def get_via_files():
        path=get_m3_animation_path()
        return glob.glob(path+'*.via')

def get_via_names():
        path=get_m3_animation_path()
        full=glob.glob(path+'*.via')
        names=[]
        for s in full:
                names.append(s[s.rfind('/')+1:s.rfind('.')])
        return names
